logprep/connector/elasticsearch/output.py

```diff
@@ -237,137 +319,158 @@ def _bulk(self, *args, **kwargs):
 
-        self._add_dates(document)
-        self.metrics.number_of_processed_events += 1
-        self._write_to_search_context(document)
+    def _handle_transport_error(self, error: search.exceptions.TransportError):
+        """Handle transport error for elasticsearch bulk indexing.
 
-    def _build_failed_index_document(self, message_document: dict, reason: str):
-        document = {
-            "reason": reason,
-            "@timestamp": TimeParser.now().isoformat(),
-            "_index": self._config.default_index,
-        }
-        try:
-            document["message"] = json.dumps(message_document)
-        except TypeError:
-            document["message"] = str(message_document)
-        return document
-
-    def store_custom(self, document: dict, target: str):
-        """Write document to Elasticsearch into the target index.
+        Discard messages that exceed the maximum size if they caused an error.
 
         Parameters
         ----------
-        document : dict
-            Document to be stored into the target index.
-        target : str
-            Index to store the document in.
-        Raises
-        ------
-        CriticalOutputError
-            Raises if any error except a BufferError occurs while writing into elasticsearch.
+        error : TransportError
+           TransportError for the error message.
 
         """
-        document["_index"] = target
-        self._add_dates(document)
-        self.metrics.number_of_processed_events += 1
-        self._write_to_search_context(document)
-
-    def store_failed(self, error_message: str, document_received: dict, document_processed: dict):
-        """Write errors into error topic for documents that failed processing.
+        if self._config.maximum_message_size_mb is None:
+            raise error
+
+        if self._is_message_exceeds_max_size_error(error):
+            messages_under_size_limit, messages_over_size_limit = self._split_list_by_size_limit()
+
+            if len(messages_over_size_limit) == 0:
+                raise error
+
+            error_documents = self._build_messages_for_large_error_documents(
+                messages_over_size_limit
+            )
+            self._message_backlog = error_documents + messages_under_size_limit
+            self._bulk(self._search_context, self._message_backlog)
+        else:
+            raise error
+
+    def _is_message_exceeds_max_size_error(self, error):
+        if error.status_code == 429:
+            if error.error == "circuit_breaking_exception":
+                return True
+
+            if error.error == "rejected_execution_exception":
+                reason = error.info.get("error", {}).get("reason", {})
+                match = self._size_error_pattern.match(reason)
+                if match and int(match.group("size")) >= int(match.group("max_size")):
+                    return True
+        return False
+
+    def _split_list_by_size_limit(self):
```

Comment:
I know this message name was my idea, but now I see that `_split_message_backlog_by_size_limit` would have been even better :D Right now it's a bit ambiguous what list will be split.